Catch Errors in Download process and only write error message

adam-james-v opened the pull request:
Fixes Private [233](https://github.com/metabase/metabase-private/issues/233)

Dissoc parts from the error that might contain viz-settings.
Adds a test confirming that viz-settings are no longer included in the error that gets written to the file.

adam-james-v committed:
walk the error to remove any visualization settings

Thanks Dan for the patch, this approach is nice because it doesn't require changing all 3 streaming-response-writer
implementations (json csv and xlsx). It keeps behaviour basically the same, but just cleans up the output to not
include potentially sensitive column/dataset details.

adam-james-v committed:
walk the error to remove any visualization settings

Thanks Dan for the patch, this approach is nice because it doesn't require changing all 3 streaming-response-writer
implementations (json csv and xlsx). It keeps behaviour basically the same, but just cleans up the output to not
include potentially sensitive column/dataset details.

adam-james-v committed:
add a test that confirms json/csv errors are cleaned

adam-james-v committed:
add a test that confirms json/csv errors are cleaned

adam-james-v committed:
Merge branch 'bugfix-dont-show-metadata-in-failed-export-233' of https://github.com/metabase/metabase into bugfix-dont-show-metadata-in-failed-export-233

adam-james-v committed:
fix dataset error test

adam-james-v committed:
Merge branch 'master' into bugfix-dont-show-metadata-in-failed-export-233

adam-james-v committed:
cljfmt

adam-james-v committed:
try to be a little more surgical with what we take out of the error

adam-james-v committed:
formatting again, oops

johnswanson commented on src/metabase/async/streaming_response.clj:
```diff
@@ -57,8 +58,15 @@
       (try
-        (with-open [writer (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))]
-          (json/generate-stream obj writer))
+        (let [obj (if (map? obj)
```
A couple thoughts here:

- nit: you don't need the outer `if` because `walk/prewalk` will apply the fn to the whole `obj` first, so you can just do 

```
(let [obj (walk/prewalk (fn [x]
                                  (if (map? x)
                                    (apply dissoc x [:visualization_settings :visualization-settings :viz-settings :preprocessed :column_settings :column-settings])
                                    x))
                                  obj)]
          (with-open [writer (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))]
            (json/generate-stream obj writer)))
```

- does this whole thing belong in `format-exception` instead of `write-error!` maybe? It feels like trying to sanitize an arbitrary map is a hard problem with only hacky solutions, whereas if we're sanitizing a Throwable it seems like we could just remove all ex-data. Something like `(update (Throwable->map e) :via (fn [cause-chain] (map #(dissoc % :data) cause-chain)))` maybe?

adam-james-v replied:
I'd say the 'hacky' part about this problem is that we're writing a stacktrace to a downloaded file at all. I think the reason we do this is to communicate *something* to the user that a download has failed, rather than giving nothing. 🤷🏼

johnswanson replied:
Sorry, "hacky" might be too harsh a way of putting it. I guess what I mean is that if `write-error!` is meant to handle an unbounded set of potential `obj` maps, it feels slightly awkward to then turn around and make assumptions about the content of those maps.

We know the current set of problematic keys, but what if someone throws a new exception somewhere that has newly problematic keys in `ex-data`? Unless there's a strong reason to give the user the ex-data, it feels simpler to just give the error messages and stacktrace and be reasonably sure we're not exposing something we shouldn't.

adam-james-v replied:
@johnswanson, I've made a change and am curious what you think. Here's some detail:

In the case of errors being saved to files, the obj is a map with a :status :failed key assoc'd by the qp somewhere
along the way. The format-exception cond branch is not used in this case, so we can't rely on modifying
`format-exception` to just dissoc ex-data.

Instead, I've kept the obj modification, but it excludes the query and the preprocessed keys, which are the only 2
keys where we see :viz-settings. This doesn't eliminate the problem of needing to exclude other keys in the future,
but it does improve the surface area - the query and viz settings are the most likely candidates for holding
potentially sensitive info.

I'd prefer to stick with this solution, as we can get a fix for the security concern in quicker, and perhaps a
non-security bug can be opened regarding the most correct way to indicate errors to users whose downloads have failed.

johnswanson replied:
Alright, that makes sense to me. Thanks!

adam-james-v committed:
pass export format, only modify the obj for file exports

In the case of errors being saved to files, the obj is a map with a :status :failed key assoc'd by the qp somewhere
along the way. The format-exception cond branch is not used in this case, so we can't rely on modifying
`format-exception` to just dissoc ex-data.

Instead, I've kept the obj modification, but it excludes the query and the preprocessed keys, which are the only 2
keys where we see :viz-settings. This doesn't eliminate the problem of needing to exclude other keys in the future,
but it does improve the surface area - the query and viz settings are the most likely candidates for holding
potentially sensitive info.

I'd prefer to stick with this solution, as we can get a fix for the security concern in quicker, and perhaps a
non-security bug can be opened regarding the most correct way to indicate errors to users whose downloads have failed.

adam-james-v committed:
keep format-exception the same

johnswanson reviewed (approved):
LGTM!

adam-james-v committed:
Merge branch 'master' into bugfix-dont-show-metadata-in-failed-export-233

adam-james-v committed:
cleaner passing of export-format to write-error!

iethree commented:
🚀 This should also be released by [v0.51](https://github.com/metabase/metabase/milestone/231)
